CONTRIBUTING.md

```diff
@@ -109,48 +109,356 @@ Check out [building instructions](https://github.com/prestodb/presto/tree/master
+* **Encourage proper encapsulation**
+    * Do not use public class members. Use lower access levels as much as possible. Exceptions may apply for performance critical paths, in that case discuss your design first with relevant code owners.
+    * Make a function access level as low as possible
+    * If a public class is only used by one caller and the usage is local, consider making it a nested class.
+* **Immutable and thread safe classes**
+    * Whenever possible, class fields should be final
+    * When it's not possible, ensure that accesses to the non-final fields are thread safe by whatever methods are appropriate for the circumstance (Concurrent collections, synchronized access, etc.) if they are in a code path that is multi-threaded
+
+* **Static imports**
+  For example, in your code, do not use something like
+
+    ```java
+    String.format("%s", rule.getClass().getSimpleName())
+    ```
+
+  * Instead, static import this method and use “format()”
+
+  ```java
+  import static java.lang.String.format;
+  ...
+  format("%s", rule.getClass().getSimpleName()) 
+  ```
+    
+  * Similarly, do NOT use
+    
+  ```java
+  time0Bto100KB.add(nanos, TimeUnit.NANOSECONDS);
+  ```
+    
+  * But static import it and use NANOSECONDS in the code.
+    
+  ```java
+  import static java.util.concurrent.TimeUnit.NANOSECONDS;
+  ...
+  time0Bto100KB.add(nanos, NANOSECONDS);
+  ```
+    
+  * Prefer Immutable collections in Guava when possible. For example, instead of using
+    
+  ```java
+  expressions.stream()
+             .map(OriginalExpressionUtils::castToExpression)
+             .collect(Collectors.toList())
+  ```
+    
+  * Use ```toImmutableList()```
+    
+  ```java
+  import com.google.common.collect.ImmutableList;
+  ...
+  import static com.google.common.collect.ImmutableList.toImmutableList;
+  ...
+  expressions.stream()
+             .map(OriginalExpressionUtils::castToExpression)
+             .collect(toImmutableList())
+  ```
+
+* **Method argument validation**
+    * Usually needed for constructors
+
+      ```java
+      protected SqlScalarFunction(Signature signature)
+      {
+          this.signature = requireNonNull(signature, "signature is null");
+          checkArgument(signature.getKind() == SCALAR, "function kind must be SCALAR");
+      }
+      ```
+
+* **Use proper annotations**
+    * @Nullable
+    * @VisibleForTesting
+    * @Override
+    * @Experimental
+* **Use of Optional instead of bare metal nullable objects when appropriate**
+    * public method parameters
+    * Performance NON-CRITICAL path
+    * Creating Optional objects is not free. If the function call is in a critical loop, do NOT use Optional parameters.
+* **Alphabetize**
+    * Sections in documentation
+    * Methods
+    * Variables
+    * Sections
+* **Comment styles**
+    * Recommend to add the “/** */” (Javadoc) style comments to all interface methods, with explanation of the parameters and returns. For example:
```

Comment:
```suggestion
    * Add the “/** */” (Javadoc) style comments to all interface methods, with explanation of the parameters and returns. For example:
```
Suggestion as shorter, and also change to imperative mode as a more visible call to action for the user. "Add" is stronger than "Recommend to add", but when would it be bad for the user to add Javadoc style comments to an interface method?